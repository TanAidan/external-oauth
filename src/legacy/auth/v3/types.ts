import { Request } from 'express'
import Integration from '../../functions/integration'
import Knex from 'knex'

export interface IAuthContext {
  clientId: string
  connectParams: any
  buid?: string
  setupId: string
  isCallback?: boolean
  store: Knex
  configuration: SetupRecord
}

interface SetupRecord {
  setup_id: string
  scopes: string[] | undefined
  buid: string
  configuration: any
}

interface IContextSessionData {
  session: {
    context: any
  }
}

export enum EDashboardApiErrorCodes {
  ENVIRONMENT_NOT_FOUND = 'ENVIRONMENT_NOT_FOUND',
  INTEGRATION_NOT_FOUND = 'INTEGRATION_NOT_FOUND'
}

export interface IDashboardApiResponse {
  errors: [
    {
      code: EDashboardApiErrorCodes
      message: string
    }
  ]
  data: {
    viewer: {
      canInvoke: boolean
      hasAccessTo: boolean
      clientId: string
      setupId: string
      buid: string
    }
  }
}

export type TConnectContextRequest = Request &
  IContextSessionData &
  IAuthContext & {
    params: {
      buid: string
    }
    query: {
      clientId: string
      setupId: string
    }
    integration?: Integration
    fullLogs?: boolean
  }

export type TCallbackContextRequest = Request & IContextSessionData & IAuthContext

export enum ELocalRedirectMethod {
  Inline = 'inline',
  Localhost = 'localhost'
}

type TNameValues = Record<string, any>

export type TIntegrationConfig = TNameValues & {
  authType: EAuthType
  config?: {
    scope?: string[]
    state?: any
  }
}

export interface IAuthConfig {
  integrationConfig: TIntegrationConfig
  setupDetails: TNameValues
}

export enum EAuthType {
  NoAuth = 'NO_AUTH',
  OAuth1 = 'OAUTH1',
  OAuth2 = 'OAUTH2'
}

interface IAuthConfigSessionData {
  session: {
    authConfig: IAuthConfig
  }
}

export type TConnectConfigRequest = Request & IAuthContext & IAuthConfig & IAuthConfigSessionData

export type TCallbackConfigRequest = Request & IAuthConfig & IAuthConfigSessionData

export interface IAuthId {
  authId: string
}

export type TAuthIdRequest = Request &
  IAuthContext &
  IAuthId & {
    session: {
      authId: string
      context: {
        setupId: string
      }
    }
  }

export interface IOAuth1Credentials {
  accessToken: string
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  expiresIn: number
}

export interface IOAuth2Credentials {
  accessToken: string
  refreshToken: string
  idToken: string
  idTokenJwt?: any
  expiresIn: number
}

export interface IWithAuthCredentials {
  credentials: IOAuth1Credentials | IOAuth2Credentials
  tokenResponse?: OAuthTokenResponse
}

export type TAuthenticateRequest = Request & IAuthContext & IAuthConfig & IWithAuthCredentials

export type AuthSuccessRequest = Request &
  IAuthContext &
  IAuthId &
  IWithAuthCredentials & { buid: string; integrationConfig?: TIntegrationConfig } & { store: Knex }

export type TRevokeRequest = Request & { buid: string }

export type TErrorHandlerRequest = Request & Partial<IAuthContext> & Partial<IAuthId>

export interface IFetchAuthDetailsParams {
  buid: string
  scopedUserDataTableName: string
  servicesTableName: string
  integration: Integration
  environmentIdentifier: string
  authId?: string
  setupId?: string
  configuration: any
  setupIdFromRequest?: boolean
  store: any
}

export interface BasicAuthDetails {
  username: string
  password: string
}

export interface ApiKeyAuthDetails {
  apiKey: string
}

export enum OAuth1SignatureMethod {
  HmacSha1 = 'HMAC-SHA1',
  RsaSha1 = 'RSA-SHA1',
  PlainText = 'PLAINTEXT'
}

interface OAuthAuthDetails {
  callbackParams: Record<string, string>
  tokenResponse?: OAuthTokenResponse
  updatedAt: number
}

export interface OAuth1AuthDetails extends OAuthAuthDetails {
  accessToken: string
  expiresIn: number
  tokenSecret: string
  consumerKey: string
  consumerSecret: string
  signatureMethod: 'HMAC-SHA1' | 'PLAINTEXT' | 'RSA-SHA1'
}

export interface OAuth2AuthDetails extends OAuthAuthDetails {
  accessToken: string
  callbackParams: Record<string, string>
  clientId: string
  clientSecret: string
  expiresIn: number
  idToken?: string
  idTokenJwt?: any
  refreshToken?: string
}

export interface OAuthTokenResponse {
  body: any
  headers: Record<string, string>
}

export type AuthDetails = BasicAuthDetails | ApiKeyAuthDetails | OAuth1AuthDetails | OAuth2AuthDetails | undefined
