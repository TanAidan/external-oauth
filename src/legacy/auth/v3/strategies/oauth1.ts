import http from 'http'
import passport from 'passport'
import { OAuth } from 'oauth'
import OAuth1Strategy from 'passport-oauth1'
import { NextFunction, Response } from 'express'
import { createNamespace } from 'cls-hooked'

import {
  TAuthenticateRequest,
  IOAuth1Credentials,
  IFetchAuthDetailsParams,
  TIntegrationConfig,
  OAuth1SignatureMethod
} from '../types'
import { AuthenticationFailed, InvalidAuthId } from '../errors'
// import { getSetupDetails, getAuth, TOAuth1Payload } from '../../../clients/integrations'
import { checkSetupIdConsistency } from './setup-id-consistency'

const responseData = createNamespace('BearerOAuth1ResponseData')

const originalCreateClient = (OAuth.prototype as any)._createClient
;(OAuth.prototype as any)._createClient = function(...args) {
  responseData.set('response', undefined)

  const request: http.ClientRequest = originalCreateClient.apply(this, args)

  request.once('response', response => {
    responseData.set('response', response)
  })

  return request
}

type VerifyCallback = (err?: Error | null, user?: object, info?: object) => void

type VerifyFunction = (
  accessToken: string,
  refreshToken: string,
  params: any,
  profile: any,
  verified: VerifyCallback
) => void

interface StrategyOptions {
  consumerKey: string
  consumerSecret: string
  requestTokenURL: string
  tokenParams?: any
  accessTokenURL: string
  userAuthorizationURL: string
  authorizationParams?: any
  signatureMethod?: string
  callbackURL: string
  customHeaders?: any
}

class Strategy extends OAuth1Strategy {
  private _authorizationParams: any
  private _tokenParams: any

  constructor(options: StrategyOptions, verify: VerifyFunction) {
    super(options, verify)
    this._authorizationParams = options.authorizationParams
    this._tokenParams = options.tokenParams
  }

  userAuthorizationParams() {
    return this._authorizationParams || {}
  }

  requestTokenParams() {
    return this._tokenParams || {}
  }

  parseErrorResponse(body: string, statusCode: number) {
    return new AuthenticationFailed({ statusCode, body })
  }
}

const strategyOptions = (req: TAuthenticateRequest) => {
  const callbackURL = process.env.AUTH_CALLBACK_URL || `${req.protocol}://${req.get('host')}/auth/callback`
  const { consumerKey, consumerSecret } = req.setupDetails.credentials
  const {
    requestTokenURL,
    tokenParams,
    accessTokenURL,
    userAuthorizationURL,
    authorizationParams,
    signatureMethod
  } = req.integrationConfig
  return {
    consumerKey,
    consumerSecret,
    requestTokenURL,
    tokenParams,
    accessTokenURL,
    userAuthorizationURL,
    authorizationParams,
    signatureMethod,
    callbackURL
  }
}

const parseExpires = (expiresIn?: string) => {
  if (!expiresIn) {
    return 0
  }

  const result = parseInt(expiresIn, 10)

  if (isNaN(result)) {
    return 0
  }

  return result
}

export const authenticate = (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  // This is invoked after we've been called back by the third party, and when Passport
  // thinks we've succesfully got a token
  const verify = (accessToken: string, tokenSecret: string, params: any, _profile: any, verified: VerifyCallback) => {
    if (!accessToken) {
      return verified(undefined, undefined, { message: 'No access token returned', response: params })
    }
    const { consumerKey, consumerSecret } = req.setupDetails

    const credentials: IOAuth1Credentials = {
      consumerKey,
      consumerSecret,
      accessToken,
      tokenSecret,
      expiresIn: parseExpires(params.expires_in)
    }

    // The OAuth library removes the token/secret from the body data.
    // Add them back in so we can store the response data as it was originally received
    const body = {
      ...params,
      oauth_token: accessToken,
      oauth_token_secret: tokenSecret
    }

    verified(undefined, { body, credentials })
  }

  // This is invoked at the end of the Passport process when either an error has occurred,
  // authentication failed, or we have successfully authenticated. This might have been invoked
  // as a consequence of the `verified` callback above, or by Passport internally
  const authenticateCallback = (err: any, data: CallbackData | undefined, info: any) => {
    if (err) {
      return next(err)
    }

    if (!data) {
      return next(AuthenticationFailed.fromOAuthRequest(req, info))
    }

    const { body, credentials } = data

    req.credentials = credentials
    req.tokenResponse = {
      body,
      headers: responseData.get('response').headers
    }

    next()
  }

  responseData.run(() => {
    passport.use(new Strategy(strategyOptions(req), verify) as any)
    passport.authenticate('oauth', authenticateCallback)(req, res, next)
  })
}

export const nullAuthDetails = {
  accessToken: null,
  tokenSecret: null,
  consumerKey: null,
  consumerSecret: null,
  signatureMethod: null
}

export const fetchAuthDetails = async (params: IFetchAuthDetailsParams, integrationConfig: TIntegrationConfig) => {
  const {
    buid,
    // servicesTableName,
    // scopedUserDataTableName,
    // environmentIdentifier,
    // integration,
    authId,
    setupIdFromRequest,
    setupId: setupIdParam
  } = params

  const { signatureMethod = OAuth1SignatureMethod.HmacSha1 } = integrationConfig

  const credentials = {
    accessToken: 'accessToken',
    callbackParamsJSON: '{}',
    connectParams: {},
    expiresIn: 1234567890,
    tokenSecret: 'tokenSecret',
    setupId: 'setupId',
    updatedAt: 1234567890,
    consumerKey: 'consumerKey',
    consumerSecret: 'consumerSecret'
  }
  // const credentials = await getAuth<TOAuth1Payload>({
  //   servicesTableName,
  //   buid: integration.buid,
  //   authId: authId!,
  //   clientId: environmentIdentifier
  // })

  if (!credentials || !credentials.accessToken) {
    throw new InvalidAuthId(buid, authId!)
  }

  const {
    accessToken,
    callbackParamsJSON,
    connectParams = {},
    expiresIn,
    tokenSecret,
    setupId,
    updatedAt
  } = credentials
  const { consumerKey, consumerSecret } = credentials

  checkSetupIdConsistency({ setupId, setupIdParam, setupIdFromRequest })

  const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

  if (!consumerKey || !consumerSecret) {
    // tslint:disable-next-line:semicolon
    // ;({ consumerKey, consumerSecret } = await getSetupDetails({
    //   scopedUserDataTableName,
    //   buid: integration.buid,
    //   setupId: setupId!,
    //   clientId: environmentIdentifier
    // }))
  }

  return {
    accessToken,
    callbackParams,
    connectParams,
    expiresIn,
    tokenSecret,
    consumerKey,
    consumerSecret,
    signatureMethod,
    updatedAt
  }
}

interface CallbackData {
  body: any
  credentials: IOAuth1Credentials
}
