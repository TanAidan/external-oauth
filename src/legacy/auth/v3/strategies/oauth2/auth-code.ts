import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getCodeRedirectURL, getTokenWithCode } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'
import { AuthenticationFailed } from '../../errors'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const callbackURL = process.env.AUTH_CALLBACK_URL || `${req.protocol}://${req.get('host')}/auth/callback`
  const {
    credentials: { clientId, clientSecret },
    scopes = []
  } = req.setupDetails
  const { code, error } = req.query
  const {
    authorizationURL,
    authorizationMethod,
    authorizationParams,
    bodyFormat,
    config,
    tokenParams,
    tokenURL
  } = req.integrationConfig

  if (error) {
    throw AuthenticationFailed.fromOAuthRequest(req, undefined)
  }

  if (code) {
    const tokenResult = await getTokenWithCode({
      authorizationMethod,
      bodyFormat,
      clientId,
      clientSecret,
      code,
      tokenParams,
      tokenURL,
      callbackURL
    })

    // console.log('tokenResult', tokenResult)
    req.credentials = responseToCredentials(tokenResult)
    // console.log('credentials', req.credentials)
    // console.log('decodedToken', tokenResult.decodedResponse)
    req.tokenResponse = tokenResult.decodedResponse

    return next()
  }

  // const { scope = [], state = 'none' } = config || {}
  const { state = 'none' } = config || {}

  const redirectURL = getCodeRedirectURL({
    authorizationParams,
    authorizationURL,
    clientId,
    state,
    scope: scopes || config?.scope || [],
    callbackURL
  })

  res.redirect(redirectURL)
})
