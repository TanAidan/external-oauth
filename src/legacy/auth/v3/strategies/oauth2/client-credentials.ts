import { NextFunction, Response } from 'express'

import { TAuthenticateRequest } from '../../types'
import { getTokenWithClientCredentials } from '../../../clients/oauth2'
import { asyncMiddleware } from '../../../../errorHandler'
import { responseToCredentials } from './common'

export const authenticate = asyncMiddleware(async (req: TAuthenticateRequest, _res: Response, next: NextFunction) => {
  const { authorizationMethod, bodyFormat, config, tokenURL } = req.integrationConfig
  const { scope = [] } = config || {}
  const { clientId, clientSecret } = req.setupDetails.credentials

  const tokenResult = await getTokenWithClientCredentials({
    authorizationMethod,
    bodyFormat,
    clientId,
    clientSecret,
    scope,
    tokenURL
  })

  req.credentials = responseToCredentials(tokenResult)
  req.tokenResponse = tokenResult.decodedResponse

  next()
})
