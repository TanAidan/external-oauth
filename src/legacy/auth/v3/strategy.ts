import { NextFunction, Response } from 'express'

import { EAuthType, TAuthenticateRequest } from './types'
import { InvalidAuthType } from './errors'
import * as oauth1 from './strategies/oauth1'
import * as oauth2 from './strategies/oauth2'
import { TBackendRequestV4 } from '../../../types'
import { asyncMiddleware } from '../../errorHandler'

const strategies = {
  [EAuthType.OAuth1]: oauth1,
  [EAuthType.OAuth2]: oauth2
}

export const isOAuthType = (authType: EAuthType) => [EAuthType.OAuth1, EAuthType.OAuth2].includes(authType)

export const authenticate = (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const { authType } = req.integrationConfig
  strategies[authType].authenticate(req, res, next)
}

export const fetchAuthDetails = asyncMiddleware(async (req: TBackendRequestV4, _res: Response, next: NextFunction) => {
  const { buid, authId, integration, configuration } = req

  const integrationConfig = await integration.config()
  const { authType } = integrationConfig

  if (!Object.values(EAuthType).includes(authType)) {
    throw new InvalidAuthType(authType)
  }

  const strategy = strategies[authType]

  const params = {
    buid,
    authId,
    integration,
    configuration,
    store: req.store
  }

  req.auth = await strategy.fetchAuthDetails(params, integrationConfig)

  console.log('[fetchAuthDetails] Auth', req.auth)

  next()
})
