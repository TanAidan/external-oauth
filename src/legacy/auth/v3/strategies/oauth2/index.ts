import { NextFunction, Response } from 'express'
// import merge from 'merge'

import { TAuthenticateRequest, IFetchAuthDetailsParams, TIntegrationConfig } from '../../types'
import { /* AccessTokenExpired,*/ InvalidGrantType, InvalidAuthId } from '../../errors'
import {
  getAuth,
  IAuthResult
  // getSetupDetails,
  // updateAuthV3
} from '../../../clients/integrations'
// import { checkSetupIdConsistency } from '../setup-id-consistency'
// import { getTokenWithRefreshToken, getTokenWithClientCredentials } from '../../../clients/oauth2'
import { authenticate as authCodeAuthenticate } from './auth-code'
import { authenticate as clientCredentialsAuthenticate } from './client-credentials'
import { /*NO_VALUE, responseToCredentials,*/ getIdTokenJwt } from './common'
// import { expandAuthConfig } from '../../../../api-config/auth-config'
// import { TOAuth2Payload } from '../../../clients/integrations'

export enum GrantType {
  AuthCode = 'authorization_code',
  ClientCredentials = 'client_credentials'
}

export const CLOCK_DRIFT_MS = 60 * 1000

export const nullAuthDetails = {
  idToken: null,
  idTokenJwt: null,
  accessToken: null
}

const strategies = {
  [GrantType.AuthCode]: authCodeAuthenticate,
  [GrantType.ClientCredentials]: clientCredentialsAuthenticate
}

export const authenticate = (req: TAuthenticateRequest, res: Response, next: NextFunction) => {
  const { grantType = GrantType.AuthCode } = req.integrationConfig

  if (!Object.values(GrantType).includes(grantType)) {
    throw new InvalidGrantType(grantType)
  }

  strategies[grantType](req, res, next)
}

// const performRefresh = async ({
//   existingCredentials,
//   integrationConfig
//   // setupDetails
// }) => {
//   const { refreshToken, idToken } = existingCredentials
//   // const { clientId, clientSecret } = setupDetails
//   const { clientId, clientSecret } = { clientId: 'clientId', clientSecret: 'clientSecret' }

//   const { authorizationMethod, bodyFormat, config, grantType, refreshURL, tokenURL } = integrationConfig

//   if (refreshToken === NO_VALUE) {
//     if (grantType === GrantType.ClientCredentials) {
//       const { scope = [] } = config || {}

//       return await getTokenWithClientCredentials({
//         authorizationMethod,
//         bodyFormat,
//         clientId,
//         clientSecret,
//         scope,
//         tokenURL
//       })
//     }

//     throw new AccessTokenExpired()
//   }

//   return await getTokenWithRefreshToken({
//     ...{ clientId: 'ClientID', clientSecret: 'clientSecret' },
//     // ...setupDetails,
//     refreshToken,
//     idToken,
//     tokenURL: refreshURL || tokenURL
//   })
// }

// const refreshAndUpdateCredentials = async (
//   params: IFetchAuthDetailsParams,
//   existingCredentials: TOAuth2Payload,
//   integrationConfig: TIntegrationConfig
// ) => {
//   const { integration, authId } = params
//   const { callbackParamsJSON, connectParams = {}, setupId, tokenResponseJSON } = existingCredentials

//   // const setupDetails = await getSetupDetails({
//   //   scopedUserDataTableName,
//   //   setupId,
//   //   buid: integration.buid,
//   //   clientId: environmentIdentifier
//   // })

//   const tokenResult = await performRefresh({
//     existingCredentials,
//     // setupDetails,
//     integrationConfig: expandAuthConfig({ connectParams, authConfig: integrationConfig as any })
//   })
//   const credentials = responseToCredentials(tokenResult)

//   const previousResponse = tokenResponseJSON ? JSON.parse(tokenResponseJSON) : {}
//   const tokenResponse = merge.recursive(previousResponse, tokenResult.decodedResponse)
//   const updatedAt = Date.now()

//   const oauthPayload: TOAuth2Payload = {
//     setupId,
//     updatedAt,
//     tokenResponseJSON: JSON.stringify(tokenResponse),
//     serviceName: integration.buid,
//     userId: authId!,
//     ...credentials
//   }

//   if (callbackParamsJSON) {
//     oauthPayload.callbackParamsJSON = callbackParamsJSON
//   }

//   // await updateAuthV3({
//   //   servicesTableName,
//   //   payload: oauthPayload,
//   //   buid: integration.buid,
//   //   authId: authId!,
//   //   clientId: environmentIdentifier
//   // })

//   const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

//   // const { clientId, clientSecret } = setupDetails
//   const { clientId, clientSecret } = { clientId: 'clientId', clientSecret: 'ClientSecret' }
//   const { accessToken, expiresIn, idToken, idTokenJwt, refreshToken } = credentials

//   return {
//     accessToken,
//     callbackParams,
//     clientId,
//     clientSecret,
//     connectParams,
//     expiresIn,
//     idToken,
//     idTokenJwt,
//     refreshToken,
//     tokenResponse,
//     updatedAt
//   }
// }

export const fetchAuthDetails = async (params: IFetchAuthDetailsParams, integrationConfig: TIntegrationConfig) => {
  const { buid, integration, authId, store, configuration } = params

  const authentication = await getAuth<IAuthResult>({
    store,
    buid: integration.buid,
    authId: authId!
  })

  if (!authentication || !authentication.payload || !authentication.payload.accessToken) {
    throw new InvalidAuthId(buid, authId!)
  }

  const {
    callbackParamsJSON,
    connectParams = {},
    expiresIn,
    updatedAt,
    accessToken,
    idToken,
    idTokenJwt,
    refreshToken,
    tokenResponseJSON
  } = authentication.payload
  const tokenResponse = tokenResponseJSON ? JSON.parse(tokenResponseJSON) : undefined
  const callbackParams = callbackParamsJSON ? JSON.parse(callbackParamsJSON) : undefined

  // checkSetupIdConsistency({ setupId, setupIdParam, setupIdFromRequest })

  const { clientId, clientSecret } = configuration.credentials
  if (!clientId || !clientSecret) {
    throw new Error('Configuration is lacking of a clientId and/or clientSecret')
  }

  // if (expiresIn) {
  //   // Sorry to be that explicit but that's something so hard to understand
  //   const expiredFromThisTime = expiresIn * 1000 + updatedAt

  //   const safeRefreshTime = expiredFromThisTime - CLOCK_DRIFT_MS

  //   const isNowLaterThanSafeRefreshTime = Date.now() > safeRefreshTime

  //   if (isNowLaterThanSafeRefreshTime) {
  //     return refreshAndUpdateCredentials(params, credentials, integrationConfig)
  //   }
  // }

  return {
    accessToken,
    callbackParams,
    connectParams,
    expiresIn,
    idToken,
    refreshToken,
    tokenResponse,
    updatedAt,
    clientId,
    clientSecret,
    idTokenJwt: idTokenJwt || getIdTokenJwt(idToken)
  }
}
