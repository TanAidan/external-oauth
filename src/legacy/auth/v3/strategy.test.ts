import * as oauth1 from './strategies/oauth1'
import * as oauth2 from './strategies/oauth2'
import { authenticate, fetchAuthDetails } from './strategy'
import { MiddlewareTestHarness } from '../../../../tests/utils'
import { TAuthenticateRequest, EAuthType } from './types'
import { RequestHandler } from 'express'
import { TBackendRequestV4 } from '../../../types'
// mport { MissingAuthId, CredentialsNotConfigured, InvalidAuthId } from './errors'

jest.mock('../../clients/integrations')
jest.mock('./strategies/api-key')
jest.mock('./strategies/basic')
jest.mock('./strategies/oauth1')
jest.mock('./strategies/oauth2')

describe('authenticate', () => {
  const testResponse = { test: 'response' }

  const setup = (authType: EAuthType) =>
    new MiddlewareTestHarness({
      configureRequest: (req: TAuthenticateRequest) => {
        req.integrationConfig = { authType }
      },
      testMiddleware: authenticate
    })

  const mockAuth = (test: MiddlewareTestHarness<TAuthenticateRequest>, middleware: RequestHandler) => {
    // mocked(middleware).mockImplementationOnce((req: TAuthenticateRequest, res: Response, next: NextFunction) => {
    //   expect(req).toBe(test.req)
    //   res.json(testResponse).end()
    //   next()
    // })
  }
  describe('when the authType is OAuth1', () => {
    it('passes through to the oauth1 authenticate middleware', async () => {
      const test = setup(EAuthType.OAuth1)
      mockAuth(test, oauth1.authenticate)

      await test
        .get()
        .expect(testResponse)
        .expect(200)
    })
  })

  describe('when the authType is OAuth2', () => {
    it('passes through to the oauth1 authenticate middleware', async () => {
      const test = setup(EAuthType.OAuth2)
      mockAuth(test, oauth2.authenticate)

      await test
        .get()
        .expect(testResponse)
        .expect(200)
    })
  })

  describe('when the authType is any other type', () => {
    it('raises an OAuthOnlyEndpoint error', async () => {
      const test = setup(EAuthType.NoAuth)

      await test.get().expect(422)

      expect(test.err).toMatchSnapshot()
    })
  })
})

describe('fetchAuthDetails', () => {
  const log = jest.fn()

  const setup = ({ withAuthId = true, withSetupId = true, authType }) =>
    new MiddlewareTestHarness({
      configureRequest: (req: TBackendRequestV4) => {
        req.integration = { buid: 'test-buid', config: jest.fn(() => ({ authType, some: 'config' })) }
        req.buid = 'test-buid'
        if (withAuthId) {
          req.authId = 'test-auth-id'
        }
        if (withSetupId) {
          req.setupId = 'test-setup-id'
        }
        req.logger = { log }
      },
      testMiddleware: fetchAuthDetails
    })

  beforeEach(() => {
    log.mockClear()
  })

  describe('when the authType is NoAuth', () => {
    it('returns no auth details', async () => {
      const test = setup({ authType: EAuthType.NoAuth })

      await test.get().expect(200)

      expect(test.req.auth).toEqual({})
    })
  })

  describe('when the authType is Basic', () => {
    const authDetails = { username: 'test-user', password: 'test-password' }

    beforeEach(() => {
      // mocked(basic.fetchAuthDetails).mockClear()
    })

    describe('when both a setupId and authId were provided', () => {
      const setupParams = { authType: EAuthType.NoAuth }

      it('uses the basic strategy to fetch the auth details', async () => {
        // mocked(basic.fetchAuthDetails).mockResolvedValueOnce(authDetails)
        const test = setup(setupParams)

        await test.get().expect(200)

        expect(test.req.auth).toEqual(authDetails)
      })
    })

    describe('when NO setupId was provided', () => {
      const setupParams = { authType: EAuthType.NoAuth, withSetupId: false }

      it('raises a CredentialsNotConfigured error', async () => {
        const test = setup(setupParams)

        await test.get().expect(422)

        expect(test.err).toMatchInlineSnapshot(`
[CredentialsNotConfigured: Missing credentials for the 'test-alias' API

Please configure the credentials in the dashboard and try again.

See the following link for information: https://docs.bearer.sh/dashboard/apis]
`)
      })
    })

    describe('when NO authId was provided', () => {
      const setupParams = { authType: EAuthType.NoAuth, withAuthId: false }

      it('uses the basic strategy to fetch the auth details', async () => {
        // mocked(basic.fetchAuthDetails).mockResolvedValueOnce(authDetails)
        const test = setup(setupParams)

        await test.get().expect(200)

        expect(test.req.auth).toEqual(authDetails)
      })
    })
  })

  describe('when the authType is ApiKey', () => {
    // const authDetails = { apiKey: 'test-api-key' }

    beforeEach(() => {
      // mocked(apiKey.fetchAuthDetails).mockClear()
    })

    describe('when both a setupId and authId were provided', () => {
      // const setupParams = { authType: EAuthType.NoAuth }

      it('uses the api key strategy to fetch the auth details', async () => {
        // mocked(apiKey.fetchAuthDetails).mockResolvedValueOnce(authDetails)
        // const test = setup(setupParams)
        // await test.get().expect(200)
        // expect(test.req.auth).toEqual(authDetails)
      })
    })

    describe('when NO setupId was provided', () => {
      const setupParams = { authType: EAuthType.NoAuth, withSetupId: false }

      it('raises a CredentialsNotConfigured error', async () => {
        const test = setup(setupParams)

        await test.get().expect(422)

        expect(test.err).toMatchInlineSnapshot(`
[CredentialsNotConfigured: Missing credentials for the 'test-alias' API

Please configure the credentials in the dashboard and try again.

See the following link for information: https://docs.bearer.sh/dashboard/apis]
`)
      })
    })

    describe('when NO authId was provided', () => {
      // const setupParams = { authType: EAuthType.NoAuth, withAuthId: false }

      it('uses the api key strategy to fetch the auth details', async () => {
        //   mocked(apiKey.fetchAuthDetails).mockResolvedValueOnce(authDetails)
        //   const test = setup(setupParams)
        //   await test.get().expect(200)
        //   expect(test.req.auth).toEqual(authDetails)
        // })
      })
    })

    describe('when the authType is OAuth1', () => {
      const authDetails = {
        accessToken: 'test-access-token',
        tokenSecret: 'test-token-secret',
        consumerKey: 'test-consumer-key',
        consumerSecret: 'test-consumer-secret'
      }

      beforeEach(() => {
        // mocked(oauth1.fetchAuthDetails).mockClear()
      })

      describe('when both a setupId and authId were provided', () => {
        const setupParams = { authType: EAuthType.OAuth1 }

        it('uses the oauth1 strategy to fetch the auth details', async () => {
          // mocked(oauth1.fetchAuthDetails).mockResolvedValueOnce(authDetails)
          const test = setup(setupParams)

          await test.get().expect(200)

          expect(test.req.auth).toEqual(authDetails)
        })
      })

      describe('when NO setupId was provided', () => {
        const setupParams = { authType: EAuthType.OAuth1, withSetupId: false }

        it('uses the oauth1 strategy to fetch the auth details', async () => {
          // mocked(oauth1.fetchAuthDetails).mockResolvedValueOnce(authDetails)
          const test = setup(setupParams)

          await test.get().expect(200)

          expect(test.req.auth).toEqual(authDetails)
        })
      })

      describe('when NO authId was provided', () => {
        const setupParams = { authType: EAuthType.OAuth1, withAuthId: false }

        it('raises a MissingAuthId error ', async () => {
          const test = setup(setupParams)

          await test.get().expect(401)

          expect(test.err).toMatchInlineSnapshot(`
[MissingAuthId: You must supply an authId to use the 'test-alias' API

Please try again with a valid authId.

See the following link for information on how to obtain an authId: https://docs.bearer.sh/faq/connect-button]
`)
        })
      })
    })

    describe('when the authType is OAuth2', () => {
      const authDetails = {
        idToken: 'test-id-token',
        accessToken: 'test-access-token'
      }

      beforeEach(() => {
        // mocked(oauth2.fetchAuthDetails).mockClear()
      })

      describe('when both a setupId and authId were provided', () => {
        const setupParams = { authType: EAuthType.OAuth2 }

        it('uses the oauth2 strategy to fetch the auth details', async () => {
          // mocked(oauth2.fetchAuthDetails).mockResolvedValueOnce(authDetails)
          const test = setup(setupParams)

          await test.get().expect(200)

          expect(test.req.auth).toEqual(authDetails)
        })
      })

      describe('when NO setupId was provided', () => {
        const setupParams = { authType: EAuthType.OAuth2, withSetupId: false }

        it('uses the oauth2 strategy to fetch the auth details', async () => {
          // mocked(oauth2.fetchAuthDetails).mockResolvedValueOnce(authDetails)
          const test = setup(setupParams)

          await test.get().expect(200)

          expect(test.req.auth).toEqual(authDetails)
        })
      })

      describe('when NO authId was provided', () => {
        const setupParams = { authType: EAuthType.OAuth2, withAuthId: false }

        it('raises a MissingAuthId error ', async () => {
          const test = setup(setupParams)

          await test.get().expect(401)

          expect(test.err).toMatchInlineSnapshot(`
[MissingAuthId: You must supply an authId to use the 'test-alias' API

Please try again with a valid authId.

See the following link for information on how to obtain an authId: https://docs.bearer.sh/faq/connect-button]
`)
        })
      })
    })

    describe('when the authType is any other value', () => {
      it('raises an InvalidAuthType error', async () => {
        const test = setup({ authType: 'invalid' })

        await test.get().expect(422)

        await expect(test.err).toMatchSnapshot()
      })
    })
  })

  it('returns null credentials when it encounters a MissingAuthId error', async () => {
    // mocked(oauth2.fetchAuthDetails).mockRejectedValueOnce(new MissingAuthId('test-alias'))
    // const test = setup()
    // await test.get().expect(200)
    // expect(test.req.auth).toEqual(oauth2.nullAuthDetails)
    // expect(log.mock.calls.map(v => v[0])).toContain(
    //   'Warning: Using null credentials as no authId parameter was specified'
    // )
  })

  it('returns null credentials when it encounters a CredentialsNotConfigured error', async () => {
    // mocked(oauth2.fetchAuthDetails).mockRejectedValueOnce(new CredentialsNotConfigured('test-alias'))
    // const test = setup()
    // await test.get().expect(200)
    // expect(test.req.auth).toEqual(oauth2.nullAuthDetails)
    // expect(log.mock.calls.map(v => v[0])).toContain('Warning: Using null credentials as no setupId was specified')
  })

  it('returns null credentials when it encounters a InvalidAuthId error', async () => {
    // mocked(oauth2.fetchAuthDetails).mockRejectedValueOnce(new InvalidAuthId('test-alias', 'test-auth-id'))
    // const test = setup()
    // await test.get().expect(200)
    // expect(test.req.auth).toEqual(oauth2.nullAuthDetails)
    // expect(log.mock.calls.map(v => v[0])).toContain(
    //   'Warning: Using null credentials as authId is not authorized for this integration'
    // )
  })

  it('fetches the auth details otherwise', async () => {
    // mocked(oauth2.fetchAuthDetails).mockResolvedValueOnce(authDetails)
    // const test = setup()
    // await test.get().expect(200)
    // expect(test.req.auth).toEqual(authDetails)
  })
})
