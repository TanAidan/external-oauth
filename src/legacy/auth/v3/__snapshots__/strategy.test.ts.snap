// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate when the authType is any other type raises an OAuthOnlyEndpoint error 1`] = `[OAuthOnlyEndpoint: This endpoint is only for 'OAUTH1' and 'OAUTH2' auth types but the integration is configured with authType 'BASIC']`;

exports[`fetchAuthDetails when the authType is any other value raises an InvalidAuthType error 1`] = `[InvalidAuthType: Invalid authType 'invalid'. Please check the integration's auth.config.json file. Allowed values are 'OAUTH1', 'OAUTH2', 'BASIC', 'APIKEY', 'NONE']`;
